### PR TITLE
Remove requirement to initialize renderer before window load

### LINKED ISSUE #14: Change direction of socket registration.
The current socket registration flow has preloader sending setup to renderer. This requires a user to call socket creation early and at the right time in renderer. This is fragile since preloader _always_ starts and loads before renderer. Any hiccup with renderer startup could result in socket setup in renderer hanging.

Instead we should have renderer talk to preloader. Since preloader is always run before renderer it is more likely and easier to ensure the preloader message receivers are setup before any renderer code is run.

### PR DESCRIPTION
**Pull request by jtbandes**
### Public-Facing Changes

Removed the requirement for the renderer to be initialized before the window's `load` event fires.

### Description

Relates to FG-2506 / https://github.com/foxglove/studio/pull/5070
Resolves https://github.com/foxglove/electron-socket/issues/14

Previously, this package required the renderer side to be initialized via `await require("@foxglove/electron-socket/renderer").Create()` **before** the window's `load` event. This was a simple mechanism for the preloader side to ensure the renderer would be present to receive its initial message containing a transferred MessagePort.

This change introduces a more complex handshake in order to allow the renderer and preload side to be initialized in any order. The new flow is:

#### If preloader is initialized first:
1. Preloader sets up a message listener.
1. Preloader sends `preloaderReady` message. The message is dropped because the renderer is not set up.
1. Renderer sets up a message listener.
1. Renderer sends a `rendererReady` message.
1. Preloader receives that message, creates a MessagePort, and sends it with a `preloaderInitialized` message.

#### If renderer is initialized first:
1. Renderer sets up a message listener.
1. Renderer sends `rendererReady` message. The message is dropped because the preloader is not set up.
1. Preloader sets up a message listener.
1. Preloader sends a `preloaderReady` message.
1. The renderer receives that message, and sends _another_ `rendererReady` message.
1. The preloader receives that message, creates a MessagePort, and sends it with a `preloaderInitialized` message.

With this change I also added a small test project that can be used to manually verify working behavior. Adding integration tests would be a good future goal.

### CODE REVIEW

File: src/preloader/PreloaderSockets.ts [hunk 1]
```diff
@@ -54,7 +55,9 @@ export class PreloaderSockets {
   ]);
 
   // A map of created `PreloaderSockets` instances
+  /** @deprecated Use `Create()` rather than accessing `registeredSockets` directly */
```
Comment (snosenzo) [inline src/preloader/PreloaderSockets.ts:58]: Why does the `PreloaderSocket` need this but the `Socket` doesn't?

**Review by snosenzo — APPROVED** [review]
can't say I really have all that deep knowledge of electron sockets, so might want some other eyes on it but the logic seems sound.

**Review by defunctzombie — APPROVED** [review]

**Comment by defunctzombie** [comment]
You can update

```
static registeredSockets = new Map<string, Sockets | Promise<Sockets>>();
```
To always store a promise to simplify reasoning about accessing items. To store a promise value you already have you can use `Promise.resolve(socket)`

**Comment by jtbandes** [comment]
Are you referring to the one in renderer/Sockets.ts? Indeed we could probably remove the `Sockets |` since it always stores a promise, or get rid of this line: `Sockets.registeredSockets.set(channel, sockets);` but it also doesn't seem like it would make a huge difference so I'll just leave it for now.

**Pull request merged**